pkg/util/vsphere/watcher/watcher.go

```diff
@@ -0,0 +1,674 @@
+
+	vm *view.Manager
+	lv *view.ListView
+	cv map[moRef]*view.ContainerView
+
+	// cvr is used to count the number of times a container has been added to
+	// the list view. If the Remove function is called on a container with a ref
+	// count of one, then the container will be removed from the list view and
+	// destroyed.
+	cvr map[moRef]map[string]struct{}
+
+	pc *property.Collector
+	pf *property.Filter
+
+	ignoredExtraConfigKeys map[string]struct{}
+	lookupNamespacedName   lookupNamespacedNameFn
+
+	err   error
+	errMu sync.RWMutex
+
+	chanDone   chan struct{}
+	chanResult chan Result
+}
+
+// Done returns a channel that is closed when the watcher is shutdown.
+func (w *Watcher) Done() <-chan struct{} {
+	return w.chanDone
+}
+
+// Result returns a channel on which new results are received.
+func (w *Watcher) Result() <-chan Result {
+	return w.chanResult
+}
+
+// Err returns the error that caused the watcher to stop.
+func (w *Watcher) Err() error {
+	w.errMu.RLock()
+	err := w.err
+	w.errMu.RUnlock()
+	return err
+}
+
+func (w *Watcher) setErr(err error) {
+	w.errMu.Lock()
+	w.err = err
+	w.errMu.Unlock()
+}
+
+func newWatcher(
+	ctx context.Context,
+	client *vim25.Client,
+	watchedPropertyPaths []string,
+	additionalIgnoredExtraConfigKeys []string,
+	lookupNamespacedName lookupNamespacedNameFn,
+	containerRefs ...moRef) (*Watcher, error) {
+
+	if watchedPropertyPaths == nil {
+		watchedPropertyPaths = DefaultWatchedPropertyPaths()
+	}
+	ignoredExtraConfigKeys := slices.Concat(
+		defaultIgnoredExtraConfigKeys,
+		additionalIgnoredExtraConfigKeys)
+
+	// Get the view manager.
+	vm := view.NewManager(client)
+
+	// For each container reference, create a container view and add it to
+	// the list view's initial list of members.
+	cvs, cvr, err := toContainerViewMap(ctx, vm, containerRefs...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a new list view used to monitor all of the containers to which
+	// VM Service VMs belong.
+	lv, err := vm.CreateListView(ctx, toMoRefs(cvs))
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a new property collector to watch for changes.
+	pc, err := property.DefaultCollector(client).Create(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a new property filter that uses the list view created up above.
+	pf, err := pc.CreateFilter(
+		ctx,
+		viewToVM(lv.Reference(), watchedPropertyPaths))
+	if err != nil {
+		return nil, err
+	}
+
+	return &Watcher{
+		chanDone:               make(chan struct{}),
+		chanResult:             make(chan Result),
+		client:                 client,
+		pc:                     pc,
+		pf:                     pf,
+		vm:                     vm,
+		lv:                     lv,
+		cv:                     cvs,
+		cvr:                    cvr,
+		ignoredExtraConfigKeys: toSet(ignoredExtraConfigKeys),
+		lookupNamespacedName:   lookupNamespacedName,
+	}, nil
+}
+
+func (w *Watcher) Close() {
+	w.once.Do(func() {
+		w.mu.Lock()
+		defer w.mu.Unlock()
```

Comment:
May not matter in this case so feel free to ignore, but in general don't want to hold a mutex lock over network calls (Destroys below).